File: java/src/main/java/com/epam/eng/pattern/flyweight/CarFactory.java
package com.epam.eng.pattern.flyweight;

import java.util.Objects;

class Car {
    private String make;
    private String model;
    private int year;
    private String color;

    public Car(String make, String model, int year, String color) {
        this.make = make;
        this.model = model;
        this.year = year;
        this.color = color;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Car car = (Car) o;
        return year == car.year && Objects.equals(make, car.make) && Objects.equals(model, car.model) && Objects.equals(color, car.color);
    }

    @Override
    public int hashCode() {
        return Objects.hash(make, model, year, color);
    }
}

public class CarFactory {
    public static Car getCar(String make, String model, int year, String color) {
        return new Car(make, model, year, color);
    }
}


// todo: use Flyweight Pattern, in CarFactory, Car objects will only be created if they do not already exist in the cache.

File: ts/flyweight/CarFactory.ts
class Car {
    private make: string;
    private model: string;
    private year: number;
    private color: string;

    constructor(make: string, model: string, year: number, color: string) {
        this.make = make;
        this.model = model;
        this.year = year;
        this.color = color;
    }

    equals(o: any): boolean {
        if (this === o) return true;
        if (o == null || this.constructor !== o.constructor) return false;
        const car: Car = o;
        return this.year === car.year && this.make === car.make && this.model === car.model && this.color === car.color;
    }

    hashCode(): number {
        return hash(this.make, this.model, this.year, this.color);
    }
}

function hash(...args: any): number {
    return args.reduce((prev: number, curr: any) => {
        return prev + (curr != null ? curr.hashCode() : 0);
    }, 0);
}

class CarFactory {
    static getCar(make: string, model: string, year: number, color: string): Car {
        return new Car(make, model, year, color);
    }
}

// todo: use Flyweight Pattern, in CarFactory, Car objects will only be created if they do not already exist in the cache.
